### ISSUE TITLE
Cannot find Python function `resource` in documentation

### ISSUE THREAD

**Issue by eschnett** [LLNL/spack#3473]
I am trying to add a `resource` call to a package. I need to know which keywords arguments it expects. I currently cannot find this in the documentation.

I notice two problems:
- When I follow a link in a search result, I end up at the top of a very large page, where I have to search again. I expect to be placed near the term which I searched.
- The two pages `spack package` and `Packaging Guide` both describe what I'm probably looking for, but they list different and incomplete keyword arguments. They also don't explain the meaning of the keyword arguments well.
- The notion of a `FetchStrategy` seems to be important to understanding resources. This term should again be a hyperlink in the API reference, so that I can look up what it means.

**Comment by adamjstewart** [comment]
> When I follow a link in a search result, I end up at the top of a very large page, where I have to search again. I expect to be placed near the term which I searched.

I don't think we have any control over that, but I agree that it is very annoying.

The rest of the comments I completely agree with.

**Comment by adamjstewart** [comment]
@scheibelp What's the difference between `destination` and `placement`? I have a resource that I need to extract in a certain directory, but when I set `destination` to that directory, it gets extracted in a subdirectory.

**Comment by adamjstewart** [comment]
Just to clarify, here's some more information on what I'm trying to do.

I'm trying to build `hoomd-blue`. It looks like the latest version requires a submodule called `libgetar`. The download tarball doesn't come with a `.git` directory, so I'm trying to expand the resource manually. The `hoomd-blue` tarball comes with an empty directory: `hoomd/extern/libgetar`. My goal is to extract the contents of the `libgetar` resource and put them in `hoomd/extern/libgetar`.

So far I have tried:
```python
destination = 'hoomd/extern/libgetar'
```
this puts the contents in a subdirectory:
```
hoomd/extern/libgetar/glotzer-libgetar-a875bd49d16c
```
```python
placement = 'hoomd/extern/libgetar'
```
I have no idea where the resource goes. I can't find it.
```python
destination = 'hoomd/extern/libgetar'
placement = 'hoomd/extern/libgetar'
```
This puts the contents in:
```
hoomd/extern/libgetar/hoomd/extern/libgetar
```
Ok, now we're getting somewhere. What about:
```python
destination = 'hoomd/extern/libgetar'
placement = '.'
```
Nope, can't find where it put it. What about:
```python
destination = '.'
placement = 'hoomd/extern/libgetar'
```
Nothing. What about:
```python
destination = 'hoomd/extern/libgetar'
placement = 'libgetar'
```
This puts the contents in:
```
hoomd/extern/libgetar/libgetar
```
Ok, we're getting close. What about:
```python
destination = 'hoomd/extern'
placement = 'libgetar'
```
Nothing. No error message, no warning, just an empty directory.

At this point I'm out of ideas. Some documentation that explains what options `resource()` takes and what they mean would be very helpful. The API docs/docstring contains:

* 'destination' : (optional) path where to move the resource. This path must be relative to the main package stage area.
* 'placement' : (optional) gives the possibility to fine tune how the resource is moved into the main package stage area.

This means nothing to me. Perhaps `placement` is supposed to be a function? I'm not sure, you can't really tell from that explanation.

**Comment by scheibelp** [comment]
This tries to explain "placement" and "destination", if this clarifies things then I'll add a doc PR for it. @adamjstewart your particular issue may require updating how resources are handled; I've marked the point in this explanation where I get to that with [ISSUE] 

Every file in a resource is moved into a location relative to the root source directory. If resource.destination is set, this relative path is augmented with resource destination. So if the root source path is /x/ and resource destination is y/z, then every resource file is moved to a location relative to /x/y/z/

Placement tells you what you want to move and where you want to move it to. If it is not set then it defaults to taking the entire expanded directory (that is, the directory itself along with the contents) and moving it to a directory with the same name as the expanded resource directory.

If you want to move the entire directory but you want to choose where it goes, then you can use placement as a string
	
```
placement="destination/path/relative/to/root/source/plus-resource.destination"
```
	
If you want to take files inside of the expanded resource directory and relocate them then you can use placement as a dictionary. For a given key-value pair, the key is the location relative to the expanded resource directory and the value is the location to move the file to (relative to the root source directory).
	
[ISSUE]
	
As of now, when the destination exists then nothing is moved, so it complicates things when the root artifact "prepares" for the addition of a resource by creating an empty directory to put it; namely this is an issue if the effect you want is to take the contents of the resource directory and place them in this prepared location.

It seems like in your case it would be useful to have an "overwrite" option that actually replaces the empty directory if it exists in the root source, or just relocates the contents of the resource directory.

Additional explanatory examples not specifically related to @adamjstewart's issue:

As an example if a resource expands to a directory "z" and you want to move that entire directory as-is into a location in the root's build path, you can set:
	
```
destination=x/y/
```

At which point after moving the resource the root source directory structure will include:
	
```
root-source/x/y/z/
```

If a resource expands to "z" and you want to call it "z-renamed" instead you can do

```
destination=x/y/
placement="z-renamed"
```

Or

```
placement="x/y/z-renamed"
```

**Comment by davydden** [comment]
@scheibelp going back to the `placement` vs ` destination` question, what if I would like to move resource into the root of `spack-src`? Can't seem to get this with any combination of `placements` and `destination`....

**Comment by adamjstewart** [comment]
This is still very poorly documented, with only a brief mention in https://spack.readthedocs.io/en/latest/packaging_guide_creation.html#extra-resources. I would like to keep this issue open until we can improve the documentation.
